crates/sui-transaction-builder/src/lib.rs

```diff
@@ -0,0 +1,1046 @@
+    fn finish(self) -> Result<Transaction, Error> {
+        let Some(sender) = self.sender else {
+            return Err(Error::NoSender);
+        };
+        if self.gas.is_empty() {
+            return Err(Error::NoGasObjects);
+        }
+        if self.gas_budget == 0 {
+            return Err(Error::NoGasBudget);
+        }
+        if self.gas_price == 0 {
+            return Err(Error::NoGasPrice);
+        }
+        if self.commands.is_empty() && !self.inputs.is_empty() {
+            return Err(Error::NoCommandsOnlyInputs);
+        }
+        if !self.commands.is_empty() && self.inputs.is_empty() {
+            return Err(Error::NoInputsOnlyCommands);
+        }
+
+        Ok(Transaction {
+            kind: sui_types::types::TransactionKind::ProgrammableTransaction(
+                sui_types::types::ProgrammableTransaction {
+                    inputs: self
+                        .inputs
+                        .into_iter()
+                        .map(|i| i.try_into())
+                        .collect::<Result<Vec<_>, _>>()
+                        .map_err(|e| {
+                            Error::InternalError(format!(
+                                "Failed to convert inputs into InputArgument: {e}"
+                            ))
+                        })?,
+                    commands: self.commands.clone(),
+                },
+            ),
+            sender,
+            gas_payment: {
+                GasPayment {
+                    objects: self
+                        .gas
+                        .clone()
+                        .into_iter()
+                        .map(|o| o.try_into())
+                        .collect::<Result<Vec<_>, _>>()
+                        .map_err(|_| {
+                            Error::InternalError(
+                                "Failed to convert gas objects into GasPayment".to_string(),
+                            )
+                        })?,
+                    owner: self.sponsor.unwrap_or(sender),
+                    price: self.gas_price,
+                    budget: self.gas_budget,
+                }
+            },
+            expiration: self.expiration.unwrap_or(TransactionExpiration::None),
+        })
+    }
+
+    /// Attempt to finish the transaction with the provided data so far.
+    ///
+    /// - if it succeeds, returns the transaction.
+    /// - if it fails, it will attempt to resolve the transaction by querying the GraphQL service
+    ///   for missing data. A client is required for this operation.
+    ///
+    /// It returns an error if the transaction cannot be resolved.
+    pub fn try_finish(self, client: Option<Client>) -> Result<Transaction, Error> {
+        match self.clone().finish() {
+            Ok(tx) => Ok(tx),
+            Err(_) => {
+                let Some(client) = client else {
+                    return Err(Error::NoClient);
+                };
+                let Some(sender) = self.sender else {
```

Comment:
rework try_finish, finish, and resolve to only have finish and resolve functions.